reconcile/status_page_components.py

```diff
@@ -0,0 +1,261 @@
+    """
+
+    id: str
+    name: str
+    description: Optional[str]
+    position: int
+    status: str
+    automation_email: Optional[str]
+    group_id: Optional[str]
+    group: Optional[bool]
+    group_name: Optional[str]
+
+
+class StatusComponent(BaseModel):
+    """
+    app-interface schema dependencies/status-page-component-1
+    """
+
+    name: str
+    display_name: str = Field(..., alias="displayName")
+    description: Optional[str]
+    group_name: Optional[str] = Field(..., alias="groupName")
+    component_id: Optional[str]
+
+
+class StatusPageProvider(BaseModel):
+    """
+    Provider specific status page reconcile implementation is implemented
+    as a subclass of `StatusPageProvider`
+    """
+
+    @abstractmethod
+    def component_ids(self) -> Iterable[str]:
+        return []
+
+    @abstractmethod
+    def apply_component(self, dry_run: bool,
+                        desired: StatusComponent) -> Optional[str]:
+        return None
+
+    @abstractmethod
+    def delete_component(self, dry_run: bool, id: str):
+        return None
+
+
+class StatusPage(BaseModel):
+    """
+    app-interface schema dependencies/status-page-1
+    """
+
+    name: str
+    page_id: str = Field(..., alias="pageId")
+    api_url: HttpUrl = Field(..., alias="apiUrl")
+    credentials: VaultSecretRef
+    components: list[StatusComponent]
+    provider: str
+
+    def get_page_provider(self) -> StatusPageProvider:
+        if self.provider == "atlassian":
+            return AtlassianStatusPage(page_id=self.page_id,
+                                       api_url=self.api_url,
+                                       token=self.credentials.get("token"))
+        else:
+            raise ValueError(f"provider {self.provider} is not supported")
+
+    def reconcile(self, dry_run: bool, state: State):
+        name_to_id_state = state.get_all("")
+        page_provider = self.get_page_provider()
+
+        # restore component ids from state
+        for desired in self.components:
+            desired.component_id = name_to_id_state.get(desired.name)
+
+        # delete
+        id_to_name_state = {v: k for k, v in name_to_id_state.items()}
+        desired_component_names = [c.name for c in self.components]
+        for current_id in page_provider.component_ids():
+            # if the component is known to the state management and if it is
+            # not known to the desired state, it was once managed by this
+            # integration but was delete from app-interface -> delete from page
+            name_for_current_component = id_to_name_state.get(current_id)
+            if (name_for_current_component and
+               name_for_current_component not in desired_component_names):
+                LOG.info(f"delete component {name_for_current_component} "
+                         f"from page {self.name}")
+                page_provider.delete_component(dry_run, current_id)
+                state.rm(name_for_current_component)
```

Comment:
Do you want to modify the state even under `dry_run`?

Reply:
oh dear good catch. i will fix this

Reply:
fixed